core/ipc/sink/common/src/main/java/org/opennms/core/ipc/sink/common/AsyncDispatcherImpl.java

```diff
@@ -150,6 +179,22 @@ public boolean offer(E e) {
 
     @Override
     public CompletableFuture<S> send(S message) {
+         
+        // Check if OffHeap is enabled and if local queue is full or if OffHeap not Empty then write message to OffHeap.
+        if (useOffHeap && (asyncPolicy.getQueueSize() == getQueueSize() ||
+                ((offHeapAdapter != null) && !offHeapAdapter.isOffHeapEmpty()))) {
+            // Start drain thread before the first write to OffHeapQueue.
+            if (offHeapAdapter == null) {
+                this.offHeapAdapter = new OffHeapAdapter();
+                offHeapAdapterExecutor.execute(offHeapAdapter);
+                LOG.info("started drain thread for {}", sinkModule.getId());
+            }
+            try {
+                return offHeapAdapter.writeMessage(message);
+            } catch (WriteFailedException e) {
+                LOG.error("OffHeap write failed ", e);
```

Comment:
If the off heap storage is full, this will flood the log with error messages and stack traces. Perhaps we can use a rate limited logger like we do with: https://github.com/OpenNMS/opennms/blob/opennms-22.0.4-1/features/newts/src/main/java/org/opennms/netmgt/newts/NewtsWriter.java#L184?